core/read.go

```diff
@@ -475,6 +476,124 @@ func readSymbol(reader *Reader, first rune) Object {
 	}
 }
 
+/* When validating symbol/keyword names, which is done only in
+   LINTER_MODE given the appropriate :rules in place, use function
+   variables for a) simplicity of functions, b) ease of adding new
+   ones (if new rules are desired), and c) hope for reasonably good
+   performance. */
+
+func isInvalidNonUnicode(s string) (bool, int, string) {
+	for i, r := range s {
+		if unicode.MaxRune < r {
+			return false, i, "not a Unicode character"
+		}
+	}
+	return true, 0, ""
+}
+
+func isInvalidNonASCII(s string) (bool, int, string) {
+	for i, r := range s {
+		if unicode.MaxASCII < r {
+			return false, i, "not an ASCII character"
+		}
+	}
+	return true, 0, ""
+}
+
+/* Returns whether a rune is a character that is inherently allowed in
+/* identifiers (symbols, keywords) by dint of the fact that
+/* clojure.core and other core packages define identifiers with these
+/* characters. While not important for parsing (Clojure is extremely
+/* permissive regarding which characters can be lexed into an
+/* identifier), linting can helpfully find and warn about characters
+/* outside of this set (as extended via configuration). */
+func isIdentifierSpecificRune(r rune) bool {
+	switch r {
+	case '*', '+', '!', '-', '?', '=', '<', '>', '&', '_', '.', '\'': // Used in clojure.core, joker.core, etc.
+		return true
+	}
+	return ('a' <= r && r <= 'z') || ('A' <= r && r <= 'Z') || ('0' <= r && r <= '9')
+}
+
+func isValidAlnum(s string) (bool, int, string) {
+	for i, r := range s {
+		if !isIdentifierSpecificRune(r) && !unicode.IsLetter(r) && !unicode.IsDigit(r) {
+			return false, i, "not a Letter nor Digit"
+		}
+	}
+	return true, 0, ""
+}
+
+func warnInvalidString(reader *Reader, validFn, invalidFn func(s string) (bool, int, string), s *string) {
+	if s == nil {
+		return
+	}
+
+	okValid, iValid, explainValid := validFn(*s)
+	okInvalid, iInvalid, explainInvalid := invalidFn(*s)
+
+	var i int
+	var explain string
+	if okValid {
+		if okInvalid {
+			return
+		}
+		i = iInvalid
+		explain = explainInvalid
+	} else {
+		if okInvalid || (iValid < iInvalid) {
+			i = iValid
+			explain = explainValid
+		} else if iValid > iInvalid {
+			i = iInvalid
+			explain = explainInvalid
+		} else { // iValid == iInvalid
+			i = iValid
+			explain = explainValid + "; " + explainInvalid
+		}
+	}
```

Comment:
The above looks more complicated than it could be... Would not it be easier to have predicate functions that check the validity of an individual rune (that it belongs to the selected character class(es) and to the selected encoding range), and then do one loop over the identifier string and report the first invalid rune? Something like 
```
func warnInvalidString(s *string) {
	for _, r := range s {
		if !isIdentSpecificRune(r) && !(isValidIdentCharacterClass(r) && isValidIdentEncodingRange(r)) {
			// print warning
			return
		}
	}
}
```

Reply:
Ignore that code you quoted; it was indeed unnecessarily complicated, mainly an attempt to just make nice diagnostics where they weren't really needed.

I'll update the PR with your proposal in the README.md, and (when I get time; very busy these next several days) provide a cleaner, and complete, implementation that I hope won't be too slow.